sextant/src/npk.rs

```diff
@@ -0,0 +1,427 @@
+    fsimg.write_all(&data_blocks.to_ne_bytes())?;
+    fsimg.write_all(&32_u16.to_ne_bytes())?;
+    fsimg.write_all(&[0_u8; 6])?;
+    fsimg.write_all(&salt.to_vec())?;
+    fsimg.write_all(&vec![0_u8; 256 - salt.len()])?;
+    fsimg.write_all(&[0_u8; 3752])?;
+    fsimg.write_all(&hash_tree)?;
+
+    Ok(())
+}
+
+fn read_signing_key(key_file_path: &Path) -> Result<SecretKey> {
+    let mut raw_signing_key_seed = [0u8; SEEDBYTES];
+    let mut sign_key_file = File::open(&key_file_path)?;
+    sign_key_file.read_exact(&mut raw_signing_key_seed)?;
+    let signing_key_seed = Seed::from_slice(&raw_signing_key_seed)
+        .with_context(|| anyhow!("Cannot read seed value"))?;
+    let (_, signing_key) = keypair_from_seed(&signing_key_seed);
+    Ok(signing_key)
+}
+
+fn gen_hashes_yaml(
+    tmp_manifest_dir: &Path,
+    fsimg_path: &Path,
+    filesystem_size: u64,
+    verity_hash: Vec<u8>,
+) -> Result<String> {
+    // create hashes YAML
+    let mut sha256 = Sha256::new();
+    io::copy(&mut File::open(&tmp_manifest_dir)?, &mut sha256)?;
+    let manifest_hash = sha256.finalize();
+    let mut sha256 = Sha256::new();
+    let mut fsimg = File::open(&fsimg_path)?;
+    io::copy(&mut fsimg, &mut sha256)?;
+
+    let fs_hash = sha256.finalize();
+    let hashes = format!(
+        "manifest.yaml:\n  hash: {:02x?}\n\
+         fs.img:\n  hash: {:02x?}\n  verity-hash: {:02x?}\n  verity-offset: {}\n",
+        manifest_hash.iter().format(""),
+        fs_hash.iter().format(""),
+        verity_hash.iter().format(""),
+        filesystem_size
+    );
+    Ok(hashes)
+}
+
+fn sign_hashes(signing_key: &SecretKey, hashes_yaml: &str) -> String {
+    let signature = sign::sign_detached(hashes_yaml.as_bytes(), &signing_key);
+    let signature_base64 = base64::encode(signature);
+    let key_id = "north";
+    let signatures_yaml = format!(
+        "{}---\nkey: {}\nsignature: {}",
+        &hashes_yaml, &key_id, &signature_base64
+    );
+    signatures_yaml
+}
+
+fn write_zip_file(
+    out_dir: &Path,
+    platform: &str,
+    manifest: &Manifest,
+    fsimg_path: &Path,
+    signatures_yaml: &str,
+) -> Result<()> {
+    let path_to_npk = out_dir
+        .join(format!(
+            "{}-{}-{}.",
+            &manifest.name, &platform, &manifest.version
+        ))
+        .with_extension("npk");
+    let npk_file = File::create(&path_to_npk)?;
+    let options =
+        zip::write::FileOptions::default().compression_method(zip::CompressionMethod::Stored);
+    let mut zip = zip::ZipWriter::new(&npk_file);
+    zip.start_file("signature.yaml", options)?;
+    zip.write_all(signatures_yaml.as_bytes())?;
+    zip.start_file("manifest.yaml", options)?;
+    let manifest_string = serde_yaml::to_string(&manifest)?;
+    zip.write_all(manifest_string.as_bytes())?;
+
+    let offset = 43 + manifest_string.len() + 44 + signatures_yaml.len() + 36; // stored
+    let padding = (offset / 4096 + 1) * 4096 - offset;
```

Comment:
same as above

Reply:
It is not clear if adding this offset is even necessary. The containers I tested work even without extra data.